[영화 극장 선택] 밀러 3, 4단계 미션 제출합니다

oungsi2000 opened the pull request:
# 셀프 체크리스트

 <!--
 모든 항목을 점검한 후 리뷰를 요청하면, 보다 의미 있는 피드백을 받을 수 있습니다.
 -->

- [x] 프로그램이 정상적으로 작동하는가?
- [x] 모든 테스트가 통과하는가?
- [x] 이전에 받은 피드백을 모두 반영하였는가?
- [x] 코딩 스타일 가이드를 준수하였는가?
    - [x] IDE 코드 자동 정렬을 적용하였는가?
    - [x] 린트 검사를 통과하였는가?
-  [x] 프로그래밍 요구 사항을 준수하였는가?
- [x] `README.md`에 기능 목록을 정리하고 명확히 기술하였는가?

# 어떤 부분에 집중하여 리뷰해야 할까요?

 <!--
 학습 목표에 맞게 구현하였는지 점검하고, 피드백을 받을 준비가 되었는지 확인하세요.
 정답을 묻기보다, 자신의 생각과 의사 결정 과정을 공유하며 피드백을 통해 논의하는 것이 좋습니다.
 -->
 
권한 설정과, db 부분에서 궁금한 점이 많습니다. 
코멘트에 궁금한 점, 어려웠던 점, 고민이 되는 점 등을 적어두었어요.
항상 감사합니다 !

## 코드 리뷰 커뮤니케이션

 <!--
 리뷰어에게 질문이 있다면, PR에 직접 코멘트를 남겨 보세요.  
 특정 코드 줄에 의견을 남기면 보다 구체적인 피드백을 받을 수 있습니다.  
 -->

📌 [GitHub에서 PR에 댓글 남기는 방법](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request)

## 참고 자료

 <!-- 관련 문서나 참고할 만한 자료를 추가하세요. -->

## 스크린샷

 <!-- 필요하다면 스크린샷을 첨부하세요. -->

시연 영상 링크입니다
https://drive.google.com/file/d/1r38nh3YJ2axyE5r5vk9AhtVVcrK4LlAr/view?usp=sharing
![]()

oungsi2000 committed:
fix: 같은 메뉴 바가 여러 번 클릭했을 때 문제 해결

같은 메뉴 아이템이 선택되면 무시하도록 변경

oungsi2000 committed:
test: 더미 데이터 추가

oungsi2000 committed:
refactor: 테스트 코드 로직 일반화

oungsi2000 committed:
feat: step3, 4 xml 파일 생성

oungsi2000 committed:
feat: ticket db로 저장

oungsi2000 committed:
feat:step3 돌아가는 쓰레기 완성

oungsi2000 committed:
feat: 알림 권한 설정 기눙 구현

oungsi2000 committed:
feat: 알림 울리는 기능 설정

oungsi2000 committed:
refactor: 데이터베이스, 프레젠터 제공자 생성

oungsi2000 committed:
refactor: 리포지토리 생성

oungsi2000 committed:
feat: 메인 스레드에서 쿼리 허용

oungsi2000 committed:
feat: 데이터 불러오는 로직 이관
presenter -> repository

oungsi2000 committed:
refactor: 리포지토리 rename

oungsi2000 committed:
refactor: SeatSelectionActivity 리포지토리 적용

oungsi2000 committed:
refactor: loadReservationInfo nullable 막음

oungsi2000 committed:
refactor: 리포지토리 구조 변경

oungsi2000 committed:
refactor: presenter by lazy로 가져오게끔 변경

oungsi2000 committed:
refactor: 상속 리포지토리 변경

oungsi2000 committed:
fix: xml 와이어프레임에 맞게 수정

oungsi2000 committed:
fix: result 뷰 컬럼이 제대로 안나오는 문제 해결'

oungsi2000 committed:
refactor: 리사이클러뷰 고도화

oungsi2000 committed:
refactor: step3 완료, 리사이클러뷰 리팩토링

oungsi2000 committed:
refactor: 패키지 구조 변경

oungsi2000 committed:
fix: setting xml 뷰 변경

oungsi2000 committed:
refactor: 지저분한 step4 코드 리팩토링

oungsi2000 committed:
feat: 알림 클릭하면 예매화면 나오는 기능 구현

oungsi2000 committed:
fix: 기존에 선택했던 선택지 유지

oungsi2000 committed:
fix: 화면 회전 시 홈 프래그먼트로 이동하는 해결

oungsi2000 committed:
test:ReservationListFragment 테스트 작성

oungsi2000 committed:
test:ReservationListPresenter 테스트 작성

데이터베이스 로드 테스트 작성

oungsi2000 committed:
test:SeatSelectionPresenter 테스트 작성

데이터베이스에 예매 정보 저장 테스트

oungsi2000 committed:
build: 의존성 충돌 해결

oungsi2000 committed:
refactor: 알람 세팅 시간 외부 주입으로 변경

oungsi2000 committed:
test:step4 테스트 작성

oungsi2000 committed:
feat: 설정 저장 기능 구현

oungsi2000 committed:
refactor: SettingFragment 데이터 바인딩 적용

oungsi2000 committed:
docs: Readme 문서 업데이트

oungsi2000 committed:
fix: 자동으로 설정 체크되게 수정

oungsi2000 committed:
fix: 권한 설정 시 자동으로 스위치 되게 수정

oungsi2000 commented on app/src/main/java/woowacourse/movie/MovieTheaterApplication.kt:
db는 싱글톤이여야 한다고 들었습니다.
때문에 커스텀 Application 진입점을 만들어 db를 초기화하도록 했는데 좋은 방법인지는 모르겠습니다.

Gyuil-Hwnag replied:
흠~ 이 부분은 Application에 대한 이해가 좀 더 필요할 것 같다는 생각이 드는군요!
Android 에서 Application이란 무엇이고, 언제 생성될까요? 추가적으로 Application의 생명주기라는 것도 존재를 하는데 이에 대해서 한번 알아본다면, Singleton과 Application에 대해서도 이해가 될 것 같다는 생각이 들었어요💪

oungsi2000 replied:
음,,,찾아보니 공식 문서에서 다음과 같이 나와있군요
>참고: 일반적으로 Application을 서브클래싱할 필요는 없습니다. 대부분의 경우 정적 싱글턴은 더욱 모듈화된 방식으로 동일한 기능을 제공할 수 있습니다. 싱글턴에 전역 컨텍스트가 필요한 경우(예: 브로드캐스트 리시버 등록), 싱글턴 메서드를 호출할 때 인수 [Context.getApplicationContext()](https://developer.android.com/reference/android/content/Context#getApplicationContext()) 로 포함하세요 . [Context](https://developer.android.com/reference/android/content/Context)getInstance()

지금처럼 단순히 db를 싱글톤으로 만들기 위해서라면 굳이 재정의할 필요가 없을 것 같다고 생각이 들었습니다 !
이 부분은 밑에서 주신 코멘트로도 이어질 것 같아요 !

추가적으로 제가 이해한 바에 따르면 시스템의 상태가 변경되었을 때, 커스터마이징 하려고 쓰는 용도 같습니다.
액티비티 / 프래그먼트와 달리
onLowMemory, registerActivityLifecycleCallbacks, onConfigurationChanged 등의 메서드가 있네요

Gyuil-Hwnag replied:
결국 Application이 사라졌군요..!🤔

oungsi2000 commented on app/src/main/java/woowacourse/movie/Provider.kt:
복잡한 의존 관계와, 뷰에서는 repository를 알게 하는게 좋지 않다고 생각하여 따로 전역 변수로 분리했습니다

Gyuil-Hwnag replied:
살펴보다보니, Provider는 서비스내에 모든 Presenter / Repository를 주입하는 형태로 되어 있군요~
이와 같은 형태로 구현이 변경된 이유가 궁금했습니다!
(Repository는 이해가 되는데 Presenter까지 추가가 된 이유가 궁금합니다)
추가적으로 만약 서비스가 커지게 된다면, Provider가 엄청 커지게 될 수도 있을 것 같다는 생각도 들었어요.

oungsi2000 replied:
3, 4단계를 구현하면서 Presenter는 repository에 의존하게 되었어요.
이전까지는 view에서 직접 presenter를 생성했지만, 이제는 직접 생성하게 된다면 view가 repository를 알게 되는 구조가 됩니다.
이를 막기 위해서 presenter 또한 Provier로 주입받는 형태로 변경했어요.

>추가적으로 만약 서비스가 커지게 된다면, Provider가 엄청 커지게 될 수도 있을 것 같다는 생각도 들었어요.

이 부분은 사실 라이브러리...를 쓰면 쉽게 해결이 될 것 같습니다 !
만약 라이브러리 없이 구현하라고 한다면, BindingAdapter를 관리한 것 처럼 view 단위로 파일을 분리하거나, 
아니면 리플렉션으로 객체를 동적으로 생성하게끔 한다면? 깔끔하게 해결될 것 같아요.

view 별로 provider를 관리하도록 변경했습니다 !

Gyuil-Hwnag replied:
테스트에 권한 관련된 Rules를 지정하는 부분도 있으니, 이부분을 살펴보는 것도 좋을 것 같군요~

oungsi2000 commented on app/src/main/java/woowacourse/movie/data/entity/TicketEntity.kt:
이 부분이 굉장히 우려스럽습니다.

현재는 하나의 테이블에 모든 예매 정보를 때려 박고 있습니다. 
특히, seats와, cinema 정보는 테이블을 분리하여 관리하고 싶은 생각이 듭니다. (제1정규형, 제3정규형)
일단 미션에서 너무 벗어나는 것 같아 하나의 테이블로 넣었습니다.

https://developer.android.com/training/data-storage/room/referencing-data?hl=ko#understand-no-object-references

해당 문서를 보면 Room은 성능 상의 이슈로 객체 간 참조를 허용하지 않고 있습니다 (다른 orm처럼 객체로 join 불가)

이로 인해 고민이 생겼는데요
1. 개발자의 명시적 join으로 인한 id값 필요 -> 도메인에서 db의 식별자인 id를 필요로 합니다 (도메인이 db를 의존하는 것아닌가?)
2. Room에서 막은 부분을 내가 구현해도 괜찮은 것인가? 

두루의 의견이 궁금합니다 !

Gyuil-Hwnag replied:
> 해당 문서를 보면 Room은 성능 상의 이슈로 객체 간 참조를 허용하지 않고 있습니다 (다른 orm처럼 객체로 join 불가)

오호 그렇군요!
이 부분에 대해서 저라면 아래와 같이 사용을 해보는 방법도 있을 것 같다는 생각이 들었어요~
join을 하지 않고, DB를 두번 탐색하여 두개의 테이블을 조회한다의 방법도 있을 것 같다는 생각이 들었어요
예시로 해본다면 아래와 같은 흐름으로 될 것 같군요!
1. reservationId를 통해서 theaterId, movieId를 가져온다.
2. theaterId로 영화관에 대한 정보를 theater테이블에서 가져온다.
3. moveId로 영화에 대한 정보를 movie테이블에서 가져온다.
4. 가져온 영화관 정보 / 영화 정보를 보여준다.

Gyuil-Hwnag replied:
만약 실제 Database에서 Join의 동작 플로우를 성능상 안되면 저희가 플로우를 실행하도록 해준다!의 느낌으로요
밀러의 생각은 어떠신가요?🤔
하나의 DB에 밀어넣는 방법도 있지만, 이렇게 한다면 개인적으로 테이블의 컬럼이 너무 무지막지하게 커지게 된다는 생각이 들었어요!

oungsi2000 replied:
음,,다시 한번 고민해보니 DB를 두번 탐색하여 두개의 테이블을 조회한다->그리고  가져온 영화관 정보 / 영화 정보를 보여준다. 라는 플로우로 간다면 결국 join을 해야 한다고 생각해요. 각각의 영화관 정보, 영화 정보를 매칭시키는 문제가 또 발생하니까요.

이렇게 생각하고, 다시 한번 고민해보니, 제가 db를 두 번 탐색하여 조합하든, Room의 @Relation 어노테이션으로 명시적 id 조인을 하든 똑같다는 생각이 들었습니다.

시간이 부족하여,,,일단 seat 테이블만 따로 빼 두었습니다만, 나중에 cinema 테이블도 분리하겠습니다
(cinema 테이블은 초기에 더미 데이터를 불러올 때, db insert를 해줘야 할 것 같아요)

이건 개인적인 추측이지만 Room이 객체 간 매핑을 직접 지원하지 않는 이유는 n+1 문제 때문이지 않을까? 하는 생각도 듭니다..!

Gyuil-Hwnag replied:
여러문제가 있을 것 같긴합니다만, 개인적으로는 성능상의 이슈가 클것 같다는 생각이 들기는 하네요!

oungsi2000 commented on app/src/main/java/woowacourse/movie/view/MainActivity.kt:
```diff
@@ -15,6 +15,18 @@ class MainActivity : BaseActivity<ActivityMainBinding>(R.layout.activity_main) {
+        }
+    }
+
+    override fun onRequestPermissionsResult(
+        requestCode: Int,
+        permissions: Array<out String>,
+        grantResults: IntArray,
+    ) {
+        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
```
이전에 구현했다가 실수로 제거하지 못했네요.
이 부분은 넘어가주셔도 될 것 같아요 !

oungsi2000 commented on app/src/main/java/woowacourse/movie/view/setting/SettingFragment.kt:
이번 step4를 구현하면서 알게 된 내용인데요 !
AlarmManager에는 크게 set, setExact로 나뉜다고 합니다.
set은 따로 권한을 필요로 하진 않지만 부정확하고, 
setExact는 Schedule_exact_time 권한이 필요합니다.

일단 step4는 시간에 관련한 것이기 때문에 setExact로 구현했어요.
하지만 set의 오차 범위를 잘 모르겠어서 set을 써도 괜찮은지 의문이 듭니다. (기준을 모르겠어요)

setExact로 구현하다보니 문제점이 이만저만이 아닌데요. 
테스트 코드에 코멘트로 남겨드리겠지만 테스트가 어렵습니다.

푸시 알림 권한 같은 경우 쉽게 권한을 설정하여 테스트가 가능하지만, 
Schedule_exact_time은 특별 권한이라 따로 라이브러리를 사용하여 진행해야 한다고 합니다...

또 자잘자잘하게 구현이 쓸데없이 복잡해진다는 느낌도 받았어요.

두루의 의견이 궁금합니다

Gyuil-Hwnag replied:
개인적으로 실제로 저도 알림 관련해서 사용을 해봤을 때, Schedule_exact_time / setExact 모두 정확한 시간을 보장하지는 못했어요..
(실제로 이전에 피처로 구현을 해봤을 때도 매번 시간을 정확하게 맞춰서 보장을 해주기에는 여러 설정 및 환경에 따라서 다르게 차이가 조금씩 존재했었네요🥲 앱 백그라운드 / 포그라운드 / 알림에 대한 중요도 / 기기설정등 여러가지 요인들이 있었습니다.)

개인적으로 이번 미션에서는 아래와 같은 내용을 키포인트로 가져가 보는 것도 좋을것 같다는 생각이 들었어요~
- 정확하지는 않더라도, 알림이 오는 로직을 구현해본다. (시간차에 대해서 연연하지 않는다.)
- Notification에 대한 구현에 초점을 맞춰본다. (정확한 알림이 오게 한다 보다는 알림 로직을 얼마나 이해를 했는지 + 어떻게 구현을 잘 분리해볼 수 있을지)

oungsi2000 commented on app/src/main/java/woowacourse/movie/view/receiver/NotificationReceiver.kt:
리시버가 너무 길어진 느낌이 드네요.

일단 isEnabled를 싱글톤(동반객체)로 설정하여 푸시 알림 설정이 꺼져있으면 모든 알림이 가지 않게(정확히는 리시버로 수신되지만 알림은 보내지 않게) 구현했어요. 

하지만 뭔가 찝찝하네요 ㅠㅠ

Gyuil-Hwnag replied:
그건 저도... 많이 찝찝하네요... (찝찝함으로 넘어서 이게 맞아? 라는 생각도 들어요🥲)
아래의 이유들로 인하여, 뭔가 구현이 복잡해졌다는 생각이 살짝 들었어요!
1. 테스트를 위해서 우선 무조건 알림을 보낸다.
2. 이러다보니, flag등에 대한 내용들도 점차 추가되었다.

그렇다면 이 방법을 해결하는 방법은 어떤 것이 있을까요?
1. testPreference등으로 이러한 flag를 대체한다. (이건 실제 로직에서도 사용이 가능하겠군요.)
2. 해당 Preference값이 true일때만 알림을 보내도록 한다.

oungsi2000 commented on app/src/androidTest/java/woowacourse/movie/view/setting/SettingFragmentTest.kt:
```diff
@@ -0,0 +1,106 @@
+    }
+
+    /**
+     * 이 테스트는 동작하지 않습니다 !!
+     * 특별 권한은 라이브러리 추가 없이 테스트가 불가능하여 일단 보류했습니다
+     * PR에 해당 내용 올리겠습니다
+     */
+    @Test
+    fun 푸시_알림이_활성화_된_경우_알림이_온다() {
+        // given
+        val manager = context.getSystemService(NotificationManager::class.java)
+        onView(withId(R.id.switch_setting_push_alarm))
+            .performClick()
+
+        // when
+        NotificationReceiver.setNotification(
+            context,
+            TestData.tickets[0],
+            LocalDateTime.now(),
+        )
+        Thread.sleep(1000)
+
+        // then
+        val result = manager.activeNotifications.find { it.id == TestData.tickets[0].hashCode() }
```
이 부분인데요. 
setExact로 하니 따로 권한 설정이 어렵네요.

그리고 channel id를 ticket의 고유값(hashcode)로 설정했는데 이게 맞는지도 모르겠습니다.

Gyuil-Hwnag replied:
개인적으로 알림 같은 경우에는 권한등과도 연관되어 있어서, 테스트가 쉽지는 않겠다는 생각이 들었어요~
필요에 의해서 라이브러리를 추가하는 방법도 있을 것 같다는 생각이 듭니다!

> channel id를 ticket의 고유값(hashcode)로 설정했는데 이게 맞는지도 모르겠습니다.

이 부분은 Notification 의 Channel Id에 대한 내용을 살펴보는 것이 좋겠다는 생각이 들었어요!
Android Notification에서 Channel Id가 의미를 하는 것이 무엇일까요?

oungsi2000 replied:
알림 채널은 안드로이드에서 푸시 알림을 카테고리별로 구별할 수 있도록 하는 객체? 인 것 같습니다. 
또 Notification의 Channel id 는 각 *카테고리*를 구별하는 역할을 한다고 합니다. (상수값)

하지만 notify의 인자로 사용되는 id는 각 알림의 식별자라고 합니다.
현재 구현의 경우 알림을 ticket 단위로 사용되고 있어 ticket의 고유값을 사용하는 것은 적절하다고 생각되네요

Gyuil-Hwnag replied:
해당 내용에 대한 의견 잘 정리해주셨네요!
스스로 이전 질문에 대한 실마리를 찾아가고 있다는 생각이 드는군요
이러한 내용들에 대해서는 같이 의논을 해보는 것도 좋긴 하지만, 아무래도 제가 코멘트를 남기기 까지에는 시간이 걸리다보니 명확하게 정의가 된 내용이다보니 관련된 내용 이해를 해본다면 사용 / 활용방법등 좀 더 자유롭게 나눠볼 아이디어들이 많아질 것 같습니다.

oungsi2000 commented on app/src/androidTest/java/woowacourse/movie/view/list/ReservationListFragmentTest.kt:
예매 내역을 불러오는 부분(db)과, 저장하는 부분, 예매 30분전에 알림을 설정하는 부분의 테스트는
전부 presenter 테스트로 이관했습니다.

Gyuil-Hwnag replied:
오호~ 좋습니다!👍
권한등에 대한 내용이 있을 수 있으니, Presenter에서 호출되는지만 테스트 해봐도 로직에 대한 검증이 충분해지겠어요

oungsi2000 reviewed (commented):
전반적으로 구현하면서 내가 구현한 방향이 맞나? 라는 생각이 들었습니다...
험하게 꾸짖어주세요.

귀중한 리뷰를 주셔서 항상 감사합니다

Gyuil-Hwnag commented on app/build.gradle.kts:
```diff
@@ -62,6 +62,8 @@ dependencies {
     implementation(libs.androidx.junit.ktx)
     implementation(libs.androidx.navigation.ui.ktx)
+    implementation(libs.room.runtime)
+    kapt(libs.androidx.room.compiler.v261)
     testImplementation(libs.assertj.core)
```
개인적으로 libs 내에 v261로 네이밍을 하는것이 맞을까? 라는 생각이 들었어요~🤔
서비스가 오래되면 될수록 라이브러리 버전들도 올라가게 될텐데, 그럼 라이브러리 버전을 업데이트 할때마다 해당 gradle을 모두 수정해야 할 것 같군요...

이렇게 지정을 하셨던 이유가 궁금합니다!

oungsi2000 replied:
음,,,사실 최신 kapt로 설정하면 빌드가 안됬습니다. 
구글링 해보니 최신 kapt가 문제가 있다는 정보를 얻어 하위 버전으로 의존성을 추가했습니다

Gyuil-Hwnag replied:
흠~ 네이밍에 대한 내용이 조금 컸었어요🥲
아니면 주석에 대한 것이라도 있었다면, 이해가 되지 않았을까? 라는 생각이 듭니다
만약 제가 실제로 일을 할때 히스트리를 모르는 상황이라면, 무조건적으로 남겼을 것 같아요

Gyuil-Hwnag commented on app/src/androidTest/java/woowacourse/movie/view/list/ReservationListFragmentTest.kt:
```diff
@@ -0,0 +1,62 @@
+@RunWith(AndroidJUnit4::class)
+@Suppress("FunctionName")
+class ReservationListFragmentTest {
+    @get:Rule
+    val scenarioRule = ActivityScenarioRule(MainActivity::class.java)
```
Fragment에 대한 테스트 전반적으로 너무 잘작성해주셨네요👍

Gyuil-Hwnag commented on app/src/androidTest/java/woowacourse/movie/view/reservation/ReservationActivityTest.kt:
```diff
@@ -59,7 +60,22 @@ class ReservationActivityTest {
+        // then
+        onView(withId(R.id.tv_reservation_count))
+            .matchText("3")
+    }
```
테스트 너무 잘해주셨는데, 만약 제가 히스토리를 모르는 상태였다면 아래와 같이 생각을 해볼 수도 있을것 같다는 생각이 들었어요!
- 초기값이 무조건1인가?

한번에 테스트 코드를 보고 알 수 있도록 아래와 같이 테스트 코드를 남겨볼 수도 있겠다는 생각도 들었어요~
테스트 코드는 로직에 대한 것을 알려주기도 하지만, 때로는 이를 보고 히스토리를 알 수 있기도 해서요!
~~~kotlin
    @Test
    fun `초기인원은_1이며_플러스버튼을_누르면_1명이_증가한다`() {
        // given
        onView(withId(R.id.tv_reservation_count))
            .matchText("1")
        
        // when
        onView(withId(R.id.btn_reservation_count_plus))
            .performClick()

        // then
        onView(withId(R.id.tv_reservation_count))
            .matchText("2")
    }
~~~

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/MovieTheaterApplication.kt:
```diff
@@ -0,0 +1,20 @@
+                "movie_theater_db",
+            )
+                // 수업 시간에 동시성, 병렬성을 배우지 않았기 때문에 일단 메인 스레드에서 쿼리를 허용하겠습니다
+                .allowMainThreadQueries()
```
이 부분은 한번 코드레빗의 코멘트도 살펴보면 좋을 것 같다는 생각이 들어서 링크를 첨부해볼게요~
(// 또는 더 간단한 방법 (기존 코드의 최소 변경) 부분 참고 부탁드릴게요!)
[코드레빗 코멘트](https://github.com/woowacourse/android-movie-theater/pull/132/files#r2083423543)

이번 미션을 통해서 Thread를 사용해보는 것도 하나의 과정일 것 같다는 생각이 듭니다!

> // 수업 시간에 동시성, 병렬성을 배우지 않았기 때문에 일단 메인 스레드에서 쿼리를 허용하겠습니다

안배웠더라도 한번 알아본다면, 다음미션등에서 좀 더 수월할 것 같다는 생각이 들었어요

oungsi2000 replied:
사실 처음에는 이렇게 서브 스레드로 db조회를 하도록 했습니다
```kotlin
thread {
    ticketDao.findAll()
...
}.join()
```
하지만 join으로 쿼리가 완료될 때까지 기다리게 된다면 결국 메인 스레드가 멈추는 것과 동일하다고 생각이 들어 그냥 메인 스레드에서 쿼리를 허용했습니다

진정한 비동기 repository 구현을 하려고 하니 두 가지 옵션이 떠올랐어요.
1. 메시지큐 방식의 비동기 처리 (쿼리가 완료되었다면 repository에서 publish 하여 메인 스레드에서 ui업데이트)
2. 코루틴

하지만 시간도 없고 오버엔지니어링이지 않을까? 싶었습니다.
이번 미션 제출 후, 코루틴으로 구현하도록 하겠습니다

Gyuil-Hwnag replied:
코루틴을 원한건 아니었어요......
거기까지 가지는 말아주세요
thread {}나 사용해주신 Transaction 어노테이션등에 대한 이해를 하길 바랬던 것이었지 코루틴(산)으로 보내고 싶은 마음은 없었답니다🥲

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/Provider.kt:
```diff
@@ -0,0 +1,43 @@
+import woowacourse.movie.view.setting.SettingContract
+import woowacourse.movie.view.setting.SettingPresenter
+
+object Provider {
+    fun ticketDao() = MovieTheaterDatabase.db.ticketDao()
```
Provider라는 네이밍을 좀 더 기능이 드러나게 한다면, 어떻게 네이밍을 지어볼 수 있을까요~?🤔

oungsi2000 replied:
각 view 별로 Provider를 분리하였습니다 !

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/Provider.kt:
```diff
@@ -0,0 +1,43 @@
+            ticketDao(),
+        )
+
+    fun settingRepository(context: Context): SettingRepository =
+        SettingRepository(
+            context.getSharedPreferences("setting", Context.MODE_PRIVATE),
+        )
```
이 부분은 한번 코드레빗의 코멘트도 살펴보면 좋을 것 같다는 생각이 들어서 링크를 첨부해볼게요~
[코드레빗 코멘트](https://github.com/woowacourse/android-movie-theater/pull/132/files#r2083423572)

oungsi2000 replied:
반영했습니다 !
다만 applicationContext 또한 Context이다보니 계속해서 파라미터로 applicationContext를 넘겨주더라도context.applicationContext를 호출해야 하는점이 불편하네요..

Gyuil-Hwnag replied:
이번 코멘트에도 한번 남겨보도록 해볼게요~
Application을 남겨두고 Manager 형태로 생성을 하고 해당 생성된 Singleton 인스턴스를 활용하는 방법이 있을 것 같았어요

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/data/MovieTheaterDatabase.kt:
```diff
@@ -0,0 +1,23 @@
+abstract class MovieTheaterDatabase : RoomDatabase() {
+    abstract fun ticketDao(): TicketDao
+
+    companion object {
+        lateinit var db: MovieTheaterDatabase
+    }
```
이 부분이 살짝 마음에 걸리긴 하군요~
factory 형태로 생성을 하는 방법도 있었을 것 같다는 생각도 들기도 하고, 아니면 Provider 에서 Context를 받아서 생성하는 방법도 있었을 것 같은데, 이와 같은 형태로 남겨두신 생각의 흐름이 궁금합니다🤔

oungsi2000 replied:
팩토리 형태로 변경했습니다 !
다만 더블 체킹 알고리즘을 넣게 되었는데...오버엔지니어링일까요..?

Gyuil-Hwnag replied:
오버엔지니어링은 아닙니다!
오히려 체크 너무 잘해주셨어요~
위에 남겨둔 것처럼 다른 방법도 있다는 것을 같이 얘기해보기 위함이었습니다
추가로 아래의 부분도 살짝 눈에 들어오기도 했었어요
~~~kotlin
        @Volatile
        private var database: MovieTheaterDatabase? = null
~~~

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/data/converter/TicketConverter.kt:
```diff
@@ -0,0 +1,43 @@
+import java.time.LocalDateTime
+import java.time.format.DateTimeFormatter
+
+class TicketConverter {
+    @TypeConverter
```
해당 Converter는 어떤 곳에서 사용을 하고 있을까요~?🤔

oungsi2000 replied:
step3을 구현하다보니 LocalDateTime과 같은 타입은 따로 변환을 해줘야 한다고 하더라구요. 
그래서 엔티티의 타입이 자바 타입, 또는 일급 컬렉션인 경우에 사용되고 있습니다 !

Gyuil-Hwnag replied:
해당 내용에 대해서 인지를 하고 잘 사용해주셨네요👍

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/data/converter/TicketConverter.kt:
```diff
@@ -0,0 +1,43 @@
+        return value.value
+    }
+
+    @TypeConverter
+    fun fromSeatList(value: List<Seat>): String {
+        return value.joinToString(",") { it.row.toString() + it.column.toString() }
+    }
```
이 부분은 코트레빗의 코멘트를 한번 살펴보는 것도 괜찮겠다는 생각이 들었어요~
필수는 아니고, 한번 참고차 입니다!
[코드레빗 코멘트](https://github.com/woowacourse/android-movie-theater/pull/132/files#r2083423566)

oungsi2000 replied:
seat을 별도의 테이블로 분리하고, 엔티티를 만들어 해당 타입 컨버터를 제거했습니다 !

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/repository/SettingRepository.kt:
```diff
@@ -0,0 +1,19 @@
+class SettingRepository(
+    private val sharedPreferences: SharedPreferences,
+) : Repository<SettingData> {
+    override fun findAll(): Result<List<SettingData>> {
+        val result = sharedPreferences.getBoolean(SettingData.NOTIFICATION_KEY, false)
+        return Result.success(listOf(SettingData(SettingData.NOTIFICATION_KEY, result)))
+    }
```
이 부분은 코드레빗의 코멘트를 한번 살펴봐도 좋을 것 같다는 생각도 들었어요~
[코드레빗 코멘트](https://github.com/woowacourse/android-movie-theater/pull/132/files#r2083423560)

추가적으로 Repository를 인터페이스로 만듬으로써 얻는 이점이 드러나지 않는 것 같다는 생각이 들기도 했습니다!
밀러는 어떤 이점을 누리기 위해서 Repository를 인터페이스로 만드셨을지가 궁금합니다!
(만약 서비스가 커져서 Repository별로 불리는 함수들이 달라지며, 기능들도 달라진다면..?🤔)

oungsi2000 replied:
음...맨 처음에 repository 인터페이스에서 자주 사용되는 메서드, findAll, findById, save, saveAll 등을 미리 구현하자! 라는 생각으로 인터페이스를 만들었는데요, Room에서는 Dao를 직접 지정해줘야 하기 때문에 결국 허투로 돌아간 것 같아요. 

하지만 딱히 문제가 되려나? 라는 생각도 들었는데 다시 생각해보니 Dao를 직접 개발자가 지정한다는 것은 repository의 다형성을 활용할 수 없다는 뜻이고, 말씀하신대로 dao별로 함수들이 달라지고 기능이 달라질 수 있겠네요.

수정했습니다 !

Gyuil-Hwnag replied:
아주 좋습니다~
Base interface에 대해서 잘못 만들다 보면 오히려 확장성을 해하는 요소가 될 수 있어서 정의할 때 주의를 하면 좋을 것 같긴 합니다💪

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/repository/TicketRepository.kt:
```diff
@@ -0,0 +1,43 @@
+        }
+    }
+
+    private fun TicketEntity.toTicket(): Ticket {
+        return Ticket(
+            title,
+            showTime,
+            seats,
+            reservationCount,
+            Cinema(1, cinemaName),
+        )
+    }
+
+    private fun Ticket.toEntity(): TicketEntity {
+        return TicketEntity(
+            title = title,
+            showTime = showTime,
+            reservationCount = reservationCount,
+            cinemaName = cinema.name,
+            seats = seats,
+            totalPrice = totalPrice(),
+        )
+    }
```
이 부분은 Mapper 형태로 따로 분리를 하게 된다면, 추후 Repository 함수가 커졌을 때에 역활별로 분리를 하고 해당 객체가 너무 뚱뚱? 해지는 것을 막을 수 있겠다는 생각도 들었어요~💪

oungsi2000 replied:
반영했습니다 !

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/view/MainActivity.kt:
```diff
@@ -24,7 +36,12 @@ class MainActivity : BaseActivity<ActivityMainBinding>(R.layout.activity_main) {
-        binding.bottomNavView.setSelectedItemId(R.id.home)
+    }
+
+    private fun setMenuBarReselectedListener() {
+        binding.bottomNavView.setOnItemReselectedListener {
+            true
+        }
```
오호 재클릭에 대한 방지를 이러한 형태로 해주셨군요👍
(다만, 이 방법은 만약 요구조건에 Fragment 내에서 뎁스가 깊어질 때 뎁스를 초기화 할 수 있도록 해주세요. 라는 요구사항이 들어간다면 까다로워질 수도 있겠군요~)
이러한 방법으로 막는 방법도 있고, 아래와 같은 방법도 있을 것 같다는 생각이 들어서 참고차 남겨두었습니다!
현재 보여지는 Fragment 가 해당 바텀시트를 클릭했을 때와 같은 Fragment 라면, Fragment의 replace를 하지 않는다. 형태로도 구현이 가능해보일 수 있겠군요~

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/view/base/BaseFragment.kt:
```diff
@@ -32,4 +39,24 @@ abstract class BaseFragment<T : ViewBinding>(val layoutId: Int) : Fragment() {
         _binding = null
     }
+
+    protected fun requestExactAlarmPermission() {
+        if (Build.VERSION.SDK_INT >= SCHEDULE_EXACT_ALARM_SDK_VERSION) {
+            startActivity(Intent(ACTION_REQUEST_SCHEDULE_EXACT_ALARM))
+        }
+    }
```
해당 기능은 모든 Fragment 에서 사용을 할 수 있도록 추가를 해주신걸까요~?🤔
이전에 한번 남겨두었던 내용같아서, 한번 참고차 남겨두도록 해볼게요!
- 밀러가 생각하시는 BaseFragment에 들어가야 한다고 생각이 드는 기능들의 기준은 무엇일까요?

oungsi2000 replied:
권한 설정은 전역에서 공유되며, 사용이 될 수 있다고 생각을 했습니다. 

현재는 settingFragment에서만 권한 요청을 하지만 만약 다른 뷰에서 권한을 사용할 일이 있다면?
또 settingFragment에서 정의한 권한 설정 및 요청 로직을 다시 작성해야 하는 번거로움이 발생할 수 있다고 생각했어요

그런 이유로 baseFragment에 넣었습니다 !

Gyuil-Hwnag replied:
흠~ 그렇군요..!
Helper등으로 뽑아내는 방법도 있었을텐데 말이지요~
자그마한 프로젝트에서도 Base에 대한 의존성이 커지는 것 같아서 눈에 계속 밟히긴 하네요

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/view/reservelist/ReservationListPresenter.kt:
```diff
@@ -0,0 +1,19 @@
+import woowacourse.movie.repository.Repository
+
+class ReservationListPresenter(
+    val view: ReservationListContract.View,
+    val repository: Repository<Ticket>,
```
해당 생성자에 대해서, open 형태로 남겨두신 이유가 궁금합니다~
코트레빗에도 비슷한 코멘트가 있었군요!
[코드 레빗 코멘트](https://github.com/woowacourse/android-movie-theater/pull/132/files#r2083423549)

oungsi2000 replied:
앗 미처 닫는걸 깜빡했네요 ! 반영했습니다

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/view/setting/SettingFragment.kt:
```diff
@@ -1,7 +1,75 @@
+    private val presenter: SettingContract.Presenter by lazy {
+        Provider.settingPresenter(this, requireContext())
+    }
+    private val requestPermissionLauncher =
+        registerForActivityResult(
+            RequestPermission(),
+        ) { isGranted: Boolean ->
+            if (hasAllPermissions()) {
+                binding.switchSettingPushAlarm.isChecked = true
+                presenter.setNotification()
+            }
+        }
```
사용자가 권한에 대한 여부를 체크한 콜백으로 설정을 하지 않고, 따로 시스템 설정을 통해서 확인을 하도록 하신 이유가 있을까요?
ActivityResult의 콜백으로 받는 내용이 의미를 하는 것이 무엇일까요~?🤔

oungsi2000 replied:
찾아보니 exact_schedule 권한과 post_notification 권한은 요청 방식이 다르더라구요. 
만약 exact_schedule권한과 post_notification권한 둘 중 하나라도 설정이 안되있다면 스위치가 켜지고 알람을 설정하는 것은 잘못된 동작이라고 생각하여 이렇게 구현했습니다 !

Gyuil-Hwnag replied:
참고차 남겨볼게요~
두개의 권한을 분리해서 체크를 하도록 남겨보는 방식이 히스토리를 모르는 사람한테는 아하~ 이래서 이렇게 했구나! 라고 더 인식이 될수도 있겠다는 생각도 들었어요
아니면 주석을 남기는 방법도 있을 것 같았어요

Gyuil-Hwnag reviewed (changes requested):
밀러 안녕하세요~
해당 미션 구현 고생 많으셨습니다!👍

사용을 하는 기능 및 내용들에 대해서 한번 이해를 해보는 시간이 되었으면 좋겠다는 생각들이 있었어요!
(한번 내용들에 대해서 간단하게라도 이해를 하고 넘어간다면, 추후에 활용방법에 대해서도 좀 더 많은 아이디어가 떠오를 것 같다는 생각이 듭니다)
추가로 고민해보시면 좋을만한 내용들 코멘트를 달았으니 확인해주시면 감사하겠습니다~

궁금한 점이나 논의하고 싶은 내용이 있으면 언제든 DM 주세요! 🙏
화이팅입니다💪

oungsi2000 committed:
refactor: 커스텀 Application 제거

oungsi2000 committed:
refactor: db thread-safe 싱글턴 구현

oungsi2000 committed:
refactor: provider 패키지 구조 변경

oungsi2000 committed:
refactor: DummyData 구조 변경

oungsi2000 committed:
refactor: Seat 테이블 분리, 및 조인으로 변경

oungsi2000 committed:
refactor: id값 int->long

oungsi2000 committed:
test: 테스트 명확성 증가

-초기값이 무조건1인가?

oungsi2000 commented:
@coderabbitai pause

oungsi2000 committed:
refactor: applicationContext 사용

oungsi2000 committed:
refactor: 안쓰는 타입 컨버터 제거

oungsi2000 committed:
refactor: Repository 인터페이스 제거

oungsi2000 committed:
refactor: Repository 인터페이스 제거

oungsi2000 committed:
refactor: domain-entity 매퍼 생성

oungsi2000 committed:
fix: 테스트 컴파일 오류 해결

oungsi2000 committed:
refactor: Presenter 프로퍼티 private 로 변경

oungsi2000 commented:
시간이 많이 부족하여 굵직굵직한 부분은 반영을 못했네요 ㅠㅠ (코루틴, setExact 제거 등)
나중에 장바구니 미션이 끝났을 때 차근차근 반영하겠습니다

Gyuil-Hwnag commented:
> 시간이 많이 부족하여 굵직굵직한 부분은 반영을 못했네요 ㅠㅠ (코루틴, setExact 제거 등) 나중에 장바구니 미션이 끝났을 때 차근차근 반영하겠습니다

코루틴을 하라고 한것은 아니었을텐데... 말이지요..ㅎㅎ😅
thread 에 대한 개념을 익혀보길 원했어요

Gyuil-Hwnag commented on app/src/androidTest/java/woowacourse/movie/view/setting/SettingFragmentTest.kt:
```diff
@@ -0,0 +1,106 @@
+    @get:Rule
+    val scenarioRule = ActivityScenarioRule(MainActivity::class.java)
+
+    @get:Rule
+    val permissionRule: GrantPermissionRule =
+        GrantPermissionRule.grant(
+            Manifest.permission.POST_NOTIFICATIONS,
+        )
```
PermissionCheck에 대한 Rule의 추가 좋네요👍
해당 내용을 활용하면, 알림 로직에 대한 검증을 해볼 수 있겠군요

Gyuil-Hwnag commented on app/src/androidTest/java/woowacourse/movie/view/setting/SettingFragmentTest.kt:
```diff
@@ -0,0 +1,106 @@
+    @Test
+    fun 푸시_알림이_비활성화_된_경우_푸시_알림이_오지_않는다() {
+        // given
+        NotificationReceiver.cancelNotification()
+        val manager = context.getSystemService(NotificationManager::class.java)
```
TestPreference를 활용하여, 검증을 해보는 방법도 있을 것 같다는 생각도 듭니다~💪

oungsi2000 replied:
현재 의존 관계가 activity(fragment) -> presenter -> repositoty -> preference 로 되어있는데, view 를 테스트하기 위해 가짜 preference를 사용하는 것은 조금 적절치 않다고 생각했습니다.

대신 cancelNotofication이 아닌 view 의 메서드 호출로 테스트 코드를 변경했습니다

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/data/MovieTheaterDatabase.kt:
```diff
@@ -0,0 +1,55 @@
+        @Volatile
+        private var database: MovieTheaterDatabase? = null
+
+        fun db(context: Context): MovieTheaterDatabase {
+            var instance = database
```
Application 에서 생성을 하는 방법도 괜찮았는데, Application이 사라지게 된 흐름이 궁금합니다~

추가로, MovieTheaterDatabase 에서 직접 인스턴스를 생성하는 방법외에도 Manager 같은 형태로 만들어두고, 해당 object Manager 내의 Database 인스턴스를 활용하도록 해보는 방법도 있을 것 같군요

oungsi2000 replied:
사실 https://developer.android.com/reference/android/app/Application 해당 문서를 보면 
>Note: There is normally no need to subclass Application. In most situations, static singletons can provide the same functionality in a more modular way. If your singleton needs a global context (for example to register broadcast receivers), include [Context.getApplicationContext()](https://developer.android.com/reference/android/content/Context#getApplicationContext()) as a [Context](https://developer.android.com/reference/android/content/Context) argument when invoking your singleton's getInstance() method.

라는 내용이 있어서 삭제하게 된 것 같아요.

>추가로, MovieTheaterDatabase 에서 직접 인스턴스를 생성하는 방법외에도 Manager 같은 형태로 만들어두고, 해당 object Manager 내의 Database 인스턴스를 활용하도록 해보는 방법도 있을 것 같군요

이 부분은 제가 잘 이해를 하지 못했는데, Provider에서 db를 반환하는 메서드도 만들었습니다 !

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/data/dao/TicketDao.kt:
```diff
@@ -0,0 +1,40 @@
+    @Insert
+    fun save(seatEntities: List<SeatEntity>)
+
+    @Transaction
+    @Query(
```
@Transaction 이라는 어노테이션은 어떤 동작을 할까요~?🤔

oungsi2000 replied:
해당 어노테이션은 db의 트랜잭션과 동일하다고 생각하여 사용했습니다
즉, 여러 개의 db 연산을 하나의 연산으로 묶어, 만약 중간에 한 쿼리가 실패했을 때 rollback 될 수 있도록 하는 역할이라고 생각했습니다.

해당 상황에서는 

val ticketId = save(ticket)
        save(seats.map { it.copy(ticketId = ticketId) }) 이 두개의 연산을 묶어주도록 하는 의도였는데 다시 보니까 위치가 잘못되었네요 ㅠㅠ 

수정했습니다 !

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/data/entity/TicketEntity.kt:
```diff
@@ -0,0 +1,17 @@
+    @PrimaryKey(autoGenerate = true) val id: Int = 0,
+    @ColumnInfo(name = "title") val title: String,
+    @ColumnInfo(name = "show_time") val showTime: LocalDateTime,
+    @ColumnInfo(name = "reservation_count") val reservationCount: ReservationCount,
+    @ColumnInfo(name = "cinema_name") val cinemaName: String,
```
Entity에서 Domain에 대해서 알고있는 것에 대한 밀러의 생각은 어떠실까요~?

oungsi2000 replied:
앗 ! 해당 부분도 엔티티로 변경하려고 했지만 시간이 부족하여 일단 Seat에만 적용했었어요!

해당 부분도 엔티티로 변경해주었습니다 !

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/provider/RepositoryProvider.kt:
```diff
@@ -0,0 +1,20 @@
+import woowacourse.movie.repository.SettingRepository
+import woowacourse.movie.repository.TicketRepository
+
+object RepositoryProvider {
+    fun ticketRepository(context: Context): TicketRepository =
```
개인적으로는 아래와 같은 형태로도 해볼 수 있겠다는 생각도 들었어요~
- ticketDao등에 대해서는 Database 관련으로 뽑아본다.

RepositoryProvider인데, Dao에 대한 구현이 있는것이 뭔가 용도에 맞게 만들었는가? 라는 생각이 드는 부분이었습니다

oungsi2000 replied:
DatabaseProvider로 해당 부분을 이동했습니다 !

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/provider/ReservationListProvider.kt:
```diff
@@ -0,0 +1,14 @@
+import woowacourse.movie.view.reservelist.ReservationListContract
+import woowacourse.movie.view.reservelist.ReservationListPresenter
+
+object ReservationListProvider {
+    fun <T> reservationListPresenter(view: T) where T : ReservationListContract.View, T : Fragment =
+        ReservationListPresenter(
+            view,
+            ticketRepository(view.requireContext().applicationContext),
+        )
+}
```
흠~ Presenter를 꼭 Provider로 받아야 하는지에 대해서는 아직까지도 계속 마음에 걸리기는 하네요🥲
밀러가 한번 저를 설득해봐주세요!!
실제로 개발을 할때에 이런 본인의 의견을 잘 피력하는 것도 중요하니, 이번 기회에 밀러의 의견을 전달하고 설득시키는 과정을 거쳐봤으면 합니다💪

oungsi2000 replied:
어제까지는 두루의 의도를 파악을 잘 하지 못했었는데 다시 생각해보니까 Application에서 가져올 수 있도록 변경했어요 !

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/view/extension/ContextExtension.kt:
```diff
@@ -0,0 +1,13 @@
+import android.app.NotificationManager
+import android.content.Context
+
+fun Context.alarmManager(): AlarmManager {
+    return getSystemService(Context.ALARM_SERVICE) as AlarmManager
+}
+
+fun Context.notificationManager(): NotificationManager {
+    return getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
+}
```
이 부분은 Extension로 만들었을 때의 장점이 어떤 것이 있었을까요~?

oungsi2000 replied:
개인적으로는 getSystemService~ 와 형변환 과정이 불필요하게 복잡해보였던 것 같아요.
조금 코틀린스럽게? 바꾼 것 같습니다

Gyuil-Hwnag reviewed (changes requested):
밀러 안녕하세요~
해당 미션 구현 고생 많으셨습니다!
리뷰 반영하면서 구현한 내용 수정 너무 잘 해주셨습니다👏

마지막 요청이다보니, 크게 고민해볼 내용들이 있을 것 같았습니다!
한번 고민한 후에 내용을 다듬는 과정에서 여러 아이디어도 나올 수 있겠다는 생각도 들었어요~
고민해보시면 좋을만한 내용들 코멘트를 달았으니 확인해주시면 감사하겠습니다.

궁금한 점이나 논의하고 싶은 내용이 있으면 언제든 DM 주세요! 🙏
마지막까지 화이팅입니다💪

oungsi2000 committed:
refactor: Cinema Join 방식으로 변경

oungsi2000 committed:
refactor: db 초기화 로직 이동

oungsi2000 committed:
refactor: 테스트 구조 수정

oungsi2000 committed:
fix: Transaction 어노테이션 위치 수정

oungsi2000 committed:
refactor: DatabaseProvider 제작

oungsi2000 committed:
refactor: CustomLiveData로 비동기 Room처리

oungsi2000 committed:
refactor: Providor 패턴 Application으로 이동

oungsi2000 committed:
refactor: CustomLiveData 제거

oungsi2000 committed:
refactor: 알람매니저 취소 로직 변경
브로드캐스트 리시버에 아예 신호를 보내지 않게끔 변경

oungsi2000 committed:
fix: 좌석 선택 시 알람 보내는 로직 변경

preference에 저장된 값 확인 후 알람 세팅,
requestcode 값을 ticket.hashcode() 로 변경하여 고유한 알람 보내게 변경

oungsi2000 committed:
refactor: repository 콜백으로 변경

oungsi2000 committed:
fix: 컴파일 안되는 에러 해결

oungsi2000 committed:
test: 알람 테스트 수정

oungsi2000 commented:
꽤나 많은 부분이 바뀌었네요. 
주요 변경 사항입니다.
1. 데이터베이스 서브 스레드 -> 콜백 방식으로 변경 
2. setExact -> set 으로 변경
3. 플래그 방식으로 알람 푸시 여부 관리 -> 알람매니저의 cancel() 메서드로 리시버로 송신 안되게끔

입니다.

항상 감사합니다

Gyuil-Hwnag commented on app/src/androidTest/java/woowacourse/movie/view/setting/SettingFragmentTest.kt:
```diff
@@ -0,0 +1,97 @@
+            .performClick()
+    }
+
+    @After
+    fun after() {
+        Intents.release()
+    }
```
@After 키워드를 통한 리소스 초기화 좋습니다!

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/MovieTheaterApplication.kt:
```diff
@@ -0,0 +1,16 @@
+import woowacourse.movie.provider.SettingProvider
+
+class MovieTheaterApplication : Application() {
+    val databaseProvider: DatabaseProvider by lazy { DatabaseProvider(this) }
+    val repositoryProvider: RepositoryProvider by lazy { RepositoryProvider(this) }
+    val reservationListProvider: ReservationListProvider by lazy { ReservationListProvider(this) }
+    val seatSelectionProvider: SeatSelectionProvider by lazy { SeatSelectionProvider(this) }
+    val settingProvider: SettingProvider by lazy { SettingProvider(this) }
```
오호~ Application의 재생성 좋습니다!👍

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/provider/DatabaseProvider.kt:
```diff
@@ -0,0 +1,22 @@
+import woowacourse.movie.repository.mapper.toEntity
+import kotlin.concurrent.thread
+
+class DatabaseProvider(context: Context) {
+    val db: MovieTheaterDatabase = MovieTheaterDatabase.db(context.applicationContext)
+    val ticketDao: TicketDao = db.ticketDao()
+
+    init {
+        val cinemaDao = db.cinemaDao()
+        thread {
+            cinemaDao.save(
+                DummyCinema.dummyCinemas.map { it.toEntity() },
+            )
+        }.join()
+    }
+}
```
Provider의 형태 처음에 생각했던 부분과 살짝 형태가 다르기는 하지만, 접근은 밀러가 구현해주신 방법과 같았어요👍

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/provider/DatabaseProvider.kt:
```diff
@@ -0,0 +1,22 @@
+
+    init {
+        val cinemaDao = db.cinemaDao()
+        thread {
+            cinemaDao.save(
+                DummyCinema.dummyCinemas.map { it.toEntity() },
+            )
+        }.join()
```
`join()` 은 어떤 동작을 할지 한번 알고 가보는 것도 도움이 될 것 같다는 생각이 들었어요~

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/provider/RepositoryProvider.kt:
```diff
@@ -0,0 +1,20 @@
+import woowacourse.movie.repository.SettingRepository
+import woowacourse.movie.repository.TicketRepository
+
+class RepositoryProvider(context: Context) {
+    private val application = context.applicationContext as MovieTheaterApplication
+
+    val ticketRepository: TicketRepository =
+        TicketRepository(
+            application.databaseProvider.ticketDao,
+        )
+
+    val settingRepository: SettingRepository =
+        SettingRepository(
+            application.getSharedPreferences("setting", Context.MODE_PRIVATE),
+        )
+}
```
처음에는 Factory 형태로 생성을 하는 것을 생각했었어요~
혹시나 나중에 시간이 남는다면, 그때 한번 도전을 해봐도 좋을 것 같다는 생각이 듭니다💪
지금 접근 너무 좋았습니다ㅎㅎ

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/repository/SettingRepository.kt:
```diff
@@ -0,0 +1,27 @@
+class SettingRepository(
+    private val sharedPreferences: SharedPreferences,
+) {
+    fun findAll(callback: (Result<List<SettingData>>) -> Unit) {
+        thread {
+            runCatching {
+                sharedPreferences.getBoolean(SettingData.NOTIFICATION_KEY, false)
+            }.onSuccess {
+                callback(Result.success(listOf(SettingData(SettingData.NOTIFICATION_KEY, it))))
+            }.onFailure {
+                callback(Result.failure(it))
+            }
+        }
+    }
```
callback 형태로 순서를 지켜주도록 하신 내용 정말 좋네요👍

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/repository/TicketRepository.kt:
```diff
@@ -0,0 +1,61 @@
+    }
+
+    fun save(
+        ticket: Ticket,
+        callback: (Result<Unit>) -> Unit,
```
Result 형태로 내보내는 방법도 있지만, onSuccess / onFailure 형태의 콜백으로 반환을 하는 것도 있을 수 있겠다는 생각이 들었어요~
아니면, runCatching 자체를 반환하는 방법도 있겠군요!

Gyuil-Hwnag commented on app/src/main/res/layout/item_reservation.xml:
```diff
@@ -0,0 +1,108 @@
+            android:reservationListTime="@{ticket.showTime}"
+            app:layout_constraintStart_toEndOf="@id/tv_line"
+            app:layout_constraintEnd_toStartOf="@id/tv_line2"
+            app:layout_constraintTop_toTopOf="@id/tv_stored_date"
+
+            tools:text="17:00" />
```
이 부분들 정렬하는 것이 빠져있었군요~
구현쪽을 유심히 보다보니 저도 xml 부분을 조금 놓쳤었군요😅

Gyuil-Hwnag commented on app/src/test/java/woowacourse/movie/presenter/setting/SettingPresenterTest.kt:
```diff
@@ -0,0 +1,46 @@
+    private lateinit var repository: TicketRepository
+    private lateinit var settingRepository: SettingRepository
+
+    @BeforeEach
+    fun setUp() {
+        view = mockk()
+        repository = mockk()
+        settingRepository = mockk()
+        presenter = SettingPresenter(view, repository, settingRepository)
+    }
```
mockk 테스트 정말 좋습니다~👍

Gyuil-Hwnag commented on gradle/libs.versions.toml:
```diff
@@ -31,7 +35,11 @@ androidx-core-ktx = { group = "androidx.core", name = "core-ktx", version.ref =
+androidx-espresso-intents = { module = "androidx.test.espresso:espresso-intents", version.ref = "androidx-espresso-core" }
 androidx-fragment-ktx = { module = "androidx.fragment:fragment-ktx", version.ref = "fragment-testing" }
+androidx-room-compiler = { module = "androidx.room:room-compiler" }
+androidx-room-compiler-v261 = { module = "androidx.room:room-compiler", version.ref = "room-compiler-version" }
+androidx-room-runtime = { module = "androidx.room:room-runtime" }
```
v261도 있지만 legacy라고 네이밍을 지어보는 것도 생각을 해봤었어요~
아니면, 해당 내용에 대한 주석이라도 있었다면 어땠을까? 라는 생각이 들었습니다!
히스토리를 모르는 사람들도 있으니, 사소한 배려차원에서요

Gyuil-Hwnag reviewed (approved):
밀러 안녕하세요~
미션 너무 고생 많으셨습니다! 👏
리뷰 반영 정말 잘 해주셨어요💯

이번 Android 미션 구현할 내용들이 많아서, 많이 어려우셨을텐데 정말 잘 마무리 해주셨어요👍
구현을 하시면서 여러가지 고민들과 키워드 들이 나왔을텐데, 밀러가 개발을 해나가시는데 조금이라도 도움이 되었으면 좋겠습니다
다음번 미션 에서도 잘 적용을 하면서 발전해 나가실 모습 기대하겠습니다😆

다음 미션 진행을 위하여 해당 미션은 여기서 머지하도록 하겠습니다.
도움이 될만한 내용들 코멘트 달았으니 확인해보셔도 좋을 것 같습니다!

궁금한 점이나 논의하고 싶은 내용이 있으면 언제든 DM 주세요! 🙏
앞으로도 응원하겠습니다 💪
